src/sentry/snuba/sessions.py

```diff
@@ -0,0 +1,167 @@
+from __future__ import absolute_import
+
+import pytz
+from datetime import datetime, timedelta
+
+from sentry.utils.snuba import raw_query, parse_snuba_datetime
+from sentry.utils.dates import to_timestamp
+from sentry.snuba.dataset import Dataset
+
+
+def _get_conditions_and_filter_keys(project_releases, environments):
+    conditions = [["release", "IN", list(x[1] for x in project_releases)]]
+    if environments is not None:
+        conditions.append(["environment", "IN", environments])
+    filter_keys = {"project_id": list(set(x[0] for x in project_releases))}
+    return conditions, filter_keys
+
+
+def get_changed_project_release_model_adoptions(project_ids):
+    """Returns the last 48 hours worth of releases."""
+    start = datetime.now(pytz.utc) - timedelta(days=2)
+    rv = []
+
+    # Find all releases with adoption in the last 24 hours
+    for x in raw_query(
+        dataset=Dataset.Sessions,
+        selected_columns=["project_id", "release", "users"],
+        groupby=["release", "project_id"],
+        start=start,
+        filter_keys={"project_id": project_ids},
+    )["data"]:
+        rv.append((x["project_id"], x["release"]))
+
+    return rv
+
+
+def get_project_releases_by_stability(project_ids, offset, limit, scope, environments=None):
+    """Given some project IDs returns adoption rates that should be updated
+    on the postgres tables.
+    """
+    assert scope in ("sessions", "users")
+
+    conditions = []
+    if environments is not None:
+        conditions.append(["environment", "IN", environments])
+    filter_keys = {"project_id": project_ids}
+    yesterday = datetime.utcnow() - timedelta(days=1)
+    rv = []
+
+    for x in raw_query(
+        dataset=Dataset.Sessions,
+        selected_columns=["project_id", "release"],
+        groupby=["release", "project_id"],
+        orderby=[["divide", ["%s_crashed" % scope, scope]]],
```

Comment:
Should you assert that `scope` is one of the valid options before sending a potentially bad query to snuba?

Reply:
There is a debug assert above.